mesonbuild/interpreter/interpreter.py

```diff
@@ -2630,8 +2629,13 @@ def source_strings_to_files(self, sources: T.List['SourceInputs']) -> T.List['So
         results: T.List['SourceOutputs'] = []
         for s in sources:
             if isinstance(s, str):
-                self.validate_within_subproject(self.subdir, s)
-                results.append(mesonlib.File.from_source_file(self.environment.source_dir, self.subdir, s))
+                if not strict and s.startswith(self.environment.get_build_dir()):
+                    results.append(s)
+                    mlog.warning(f'Source item {s!r} cannot be converted to File object, because it is a generated file. '
+                                 'This will become a hard error in the future.', location=self.current_node)
+                else:
+                    self.validate_within_subproject(self.subdir, s)
+                    results.append(mesonlib.File.from_source_file(self.environment.source_dir, self.subdir, s))
```

Comment:
I don't see how? We pass through mesonlib.File already, this check only covers `isinstance(s, str)`. and explicitly uses `from_source_file()` which sets `File(is_built: False, ...)` and raises MesonException if the file does not exist *in the source dir*.

So it should be impossible for this change to add a warning, unless *this* function used to raise a fatal error.

And, the kicker is that it only successfully works for custom_target because we try/except and ignore MesonException, and then print a warning anyway -- so all I did was move the warning from one place to another. Unless there is some other location where we ignore exceptions?

Reply:
A quick check tells me that nowhere else is `source_strings_to_files` invoked in a try/except. And the one place where it is invoked as such, we always raised a similar warning, which I have narrowed to the case where a path to the build directory is used.

Reply:
To be fair, if we ever use an *absolute path* as a string, from_source_file() will successfully find the result anyway because `os.path.join(abspath, abspath)`.

But I would be very surprised if we ever pass string types around like this for generated files, and instead of using `File.from_built_file()` used `os.path.join(self.environment.build_dir, generated_filename_string)`.

Reply:
RIght, okay. Then yeah, this is fine then. Since it is only checking strings.